src/mds/OpenFileTable.cc

```diff
@@ -1053,35 +1053,33 @@ void OpenFileTable::_prefetch_dirfrags()
   ceph_assert(prefetch_state == DIRFRAGS);
 
   MDCache *mdcache = mds->mdcache;
-  std::vector<CDir*> fetch_queue;
+  std::set<CDir*> fetch_queue;
 
-  for (auto& it : loaded_anchor_map) {
-    if (it.second.frags.empty())
+  for (auto& [ino, anchor] : loaded_anchor_map) {
+    if (anchor.frags.empty())
       continue;
-    CInode *diri = mdcache->get_inode(it.first);
+    CInode *diri = mdcache->get_inode(ino);
     if (!diri)
       continue;
     if (diri->state_test(CInode::STATE_REJOINUNDEF))
       continue;
 
-    for (auto& fg: it.second.frags) {
+    for (auto& fg: anchor.frags) {
       CDir *dir = diri->get_dirfrag(fg);
       if (dir) {
 	if (dir->is_auth() && !dir->is_complete())
-	  fetch_queue.push_back(dir);
+	  fetch_queue.insert(dir);
       } else {
 	frag_vec_t leaves;
 	diri->dirfragtree.get_leaves_under(fg, leaves);
-	if (leaves.empty())
-	  leaves.push_back(diri->dirfragtree[fg.value()]);
```

Comment:
If this change should fix the issue, we can get rid of the `std::set` conversion above. Vectors are the preferred data structure.

Reply:
Yeah, after dirfragments merged the leaves vector will be empty when trying to get the leaves under any of that stale sub directors. And then this two lines will push the merged new CDir always.

Since the merged new CDir should already in the mds cache, no need to prefech it here, so removing them should be okay.